usecases/modulecomponents/batch/batch.go

```diff
@@ -124,108 +159,210 @@ func (b *Batch) batchWorker() {
-						if !job.skipObject[j] {
-							job.errs[j] = errors.New("request rate limit exceeded and will not refresh in time")
-						}
+		// in case of low rate limits we should not send the next batch immediately but sleep a bit
+		batchesPerMinute := 61.0 / batchTookInS
+		if batchesPerMinute > float64(rateLimit.LimitRequests) {
+			sleepFor := time.Duration((60.0-batchTookInS*float64(rateLimit.LimitRequests))/float64(rateLimit.LimitRequests)) * time.Second
+			time.Sleep(sleepFor)
+
+			// adapt the batches per limit
+			batchesPerMinute = float64(rateLimit.LimitRequests)
+		}
+		if batchesPerMinute*float64(tokensInCurrentBatch) > float64(rateLimit.LimitTokens) {
+			sleepFor := batchTookInS * (batchesPerMinute*float64(tokensInCurrentBatch) - float64(rateLimit.LimitTokens)) / float64(rateLimit.LimitTokens)
+			time.Sleep(time.Duration(sleepFor * float64(time.Second)))
+		}
+
+		// not all request limits are included in "RemainingRequests" and "ResetRequests". For example, in the OPenAI
+		// free tier only the RPD limits are shown but not RPM
+		if rateLimit.RemainingRequests <= 0 && time.Until(rateLimit.ResetRequests) > 0 {
+			// if we need to wait more than MaxBatchTime for a reset we need to stop the batch to not produce timeouts
+			if time.Since(job.startTime)+time.Until(rateLimit.ResetRequests) > b.maxBatchTime {
+				for j := origIndex[0]; j < len(job.texts); j++ {
+					if !job.skipObject[j] {
+						job.errs[j] = errors.New("request rate limit exceeded and will not refresh in time")
 					}
-					break
 				}
-				time.Sleep(time.Until(rateLimit.ResetRequests))
+				break
 			}
-
-			// reset for next vectorizer-batch
-			tokensInCurrentBatch = 0
-			texts = texts[:0]
-			origIndex = origIndex[:0]
+			time.Sleep(time.Until(rateLimit.ResetRequests))
 		}
 
-		// in case we exit the loop without sending the last batch. This can happen when the last object is a skip or
-		// is too long
-		if len(texts) > 0 && objCounter == len(job.texts) {
-			_ = b.makeRequest(job, texts, job.cfg, origIndex, rateLimit, tokensInCurrentBatch)
-		}
-		rateLimitPerApiKey[job.apiKeyHash] = rateLimit
-		job.wg.Done()
+		// reset for next vectorizer-batch
+		tokensInCurrentBatch = 0
+		texts = texts[:0]
+		origIndex = origIndex[:0]
 	}
+
+	// in case we exit the loop without sending the last batch. This can happen when the last object is a skip or
+	// is too long
+	if len(texts) > 0 && objCounter == len(job.texts) {
+		_ = b.makeRequest(job, texts, job.cfg, origIndex, rateLimit, tokensInCurrentBatch)
+	}
+	objectsPerRequest := 0
+	if numRequests > 0 {
+		objectsPerRequest = numSendObjets / numRequests
+	}
+	b.endOfBatchChannel <- endOfBatchJob{
+		timePerToken:      timePerToken,
+		objectsPerRequest: objectsPerRequest,
+		reservedTokens:    job.tokenSum,
+		reservedReqs:      reservedReqs,
+		apiKeyHash:        job.apiKeyHash,
+		concurrentBatch:   concurrentBatch,
+	}
+	job.wg.Done()
+	b.concurrentBatches.Add(-1)
```

Comment:
can this be moved outside of the method. it would be easier to understand what this variable is used for if all the writes and reads were in 1 place.

Reply:
In concurrent mode we can have multiple goroutines running with this code. We want to decrement at the end of the function, when the batch is done so we always have a correct view on how many gorutines are currently running.

If anything is unclear I am happy to explain more or add more comments